File: pydroid-ui/src/main/java/com/pyamsoft/pydroid/ui/PYDroid.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.pydroid.ui;

import android.content.Context;
import android.os.StrictMode;
import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.annotation.VisibleForTesting;
import com.pyamsoft.pydroid.PYDroidModule;
import com.pyamsoft.pydroid.helper.Checker;
import timber.log.Timber;

public final class PYDroid {

  @Nullable private static volatile PYDroid instance = null;
  @NonNull private final PYDroidComponent component;
  private final boolean debug;

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  @SuppressWarnings("WeakerAccess") PYDroid(@NonNull PYDroidModule module) {
    module = Checker.checkNonNull(module);
    component = PYDroidComponentImpl.withModule(module);
    debug = module.isDebug();

    UiLicenses.addLicenses();
    if (module.isDebug()) {
      setStrictMode();
      Timber.plant(new Timber.DebugTree());
    }

    Timber.i("Initialize PYDroid Injector singleton");
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  @NonNull @CheckResult public static PYDroid getInstance() {
    if (instance == null) {
      synchronized (PYDroid.class) {
        if (instance == null) {
          throw new IllegalStateException(
              "PYDroid instance is NULL, create it first using PYDroid.initialize(Context, boolean)");
        }
      }
    }
    return Checker.checkNonNull(instance);
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  @VisibleForTesting
  public static void setInstance(@NonNull PYDroid pyDroid) {
    synchronized (PYDroid.class) {
      Timber.w("Manually settings PYDroid instance");
      instance = pyDroid;
    }
  }

  /**
   * Initialize the library
   */
  public static void initialize(@NonNull Context context, boolean debug) {
    if (instance == null) {
      synchronized (PYDroid.class) {
        if (instance == null) {
          instance = new PYDroid(new PYDroidModule(context.getApplicationContext(), debug));
        }
      }
    }

    if (instance == null) {
      throw new RuntimeException("PYDroid initialization failed!");
    }
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  @CheckResult
  public boolean isDebugMode() {
    return debug;
  }

  @RestrictTo(RestrictTo.Scope.LIBRARY)
  @CheckResult @NonNull
  public PYDroidComponent provideComponent() {
    return Checker.checkNonNull(component);
  }

  /**
   * Sets strict mode flags when running in debug mode
   */
  private void setStrictMode() {
    StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder().detectAll()
        .penaltyLog()
        .penaltyDeath()
        .permitDiskReads()
        .permitDiskWrites()
        .penaltyFlashScreen()
        .build());
    StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder().detectAll().penaltyLog().build());
  }
}

File: pydroid-ui/src/main/java/com/pyamsoft/pydroid/ui/PYDroid.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.ui

import android.app.Application
import android.content.Context
import android.os.StrictMode
import android.support.annotation.CheckResult
import com.pyamsoft.pydroid.loader.LoaderModule
import com.pyamsoft.pydroid.ui.about.UiLicenses
import timber.log.Timber

/**
 * PYDroid library entry point
 */
class PYDroid private constructor(
  application: Application,
  debug: Boolean
) {

  private val component: PYDroidComponent = PYDroidComponentImpl(application, debug)

  init {
    if (debug) {
      Timber.plant(Timber.DebugTree())
      setStrictMode()
    }
    UiLicenses.addLicenses()
  }

  /**
   * Exposed so that outside applications can take advantage of the ImageLoader instance and cache
   */
  @CheckResult
  fun loaderModule(): LoaderModule {
    return component.loaderModule()
  }

  interface Instance {

    @CheckResult
    fun getPydroid(): PYDroid?

    fun setPydroid(instance: PYDroid)

  }

  companion object {

    @JvmStatic
    private fun setStrictMode() {
      StrictMode.setThreadPolicy(
          StrictMode.ThreadPolicy.Builder()
              .detectAll()
              .penaltyLog()
              .penaltyDeath()
              .permitDiskReads()
              .permitDiskWrites()
              .penaltyFlashScreen()
              .build()
      )
      StrictMode.setVmPolicy(
          StrictMode.VmPolicy.Builder()
              .detectAll()
              .penaltyLog()
              .build()
      )
    }

    /**
     * Access point for library component graph
     */
    @JvmStatic
    @CheckResult
    internal fun obtain(context: Context): PYDroidComponent {
      val app = context.applicationContext
      if (app is PYDroid.Instance) {
        return checkNotNull(app.getPydroid()?.component) {
          "PYDroid not initialized. call PYDroid.init(Application, PYDroid.Instance, Boolean)"
        }
      } else {
        throw IllegalStateException("Application does not implement PYDroid.Instance interface")
      }
    }

    /**
     * Initialize the library
     *
     * You should carry the passed modules with you to any other component graphs or you will have "doubled" singletons
     */
    @JvmStatic
    fun init(
      application: Application,
      instance: Instance,
      debug: Boolean
    ) {
      if (instance.getPydroid() == null) {
        instance.setPydroid(PYDroid(application, debug))
      }
    }
  }
}
